include/dlaf/common/callable_object.h

```diff
@@ -0,0 +1,27 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2020, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// Given a function name @fname generates a constexpr object with name fname_o
+/// with type fname_t. The generated type has one static operator() which
+/// transparently forwards all arguments to a call to fname. This macro is
+/// useful for creating wrappers of overloaded functions that would otherwise
+/// require explicit selection of the overload when passing the function to
+/// higher-order functions. With this wrapper the overload is selected inside
+/// the wrapper's operator().
+#define DLAF_MAKE_CALLABLE_OBJECT(fname)     \
+  constexpr struct fname##_t {               \
+    template <typename... Ts>                \
+    decltype(auto) operator()(Ts&&... ts) {  \
```

Comment:
I was playing with something similar...

wouldn't it be betteer to have it `const noexcept`?

Reply:
Good point! I think it should be `const` and conditionally `noexcept`. I'll leave this open to remind me to change this.

Reply:
Also, as discussed earlier with @albestro an alternative to this would be something like this `LIFT` macro: https://github.com/rollbear/lift/blob/42293d66f0bd5a92b2a0ab42284e56950d56d853/include/lift.hpp#L20-L32, which uses a lambda instead of creating a callable object. It's less boilerplate when defining functions, but requires a `LIFT(fun)` when used.

Reply:
This is now updated to be const and conditionally noexcept.